lgc/patch/PatchInOutImportExport.cpp

```diff
@@ -5521,6 +5525,328 @@ Value *PatchInOutImportExport::swizzleLocalInvocationIdIn8x4(Value *localInvocat
+  //   // Calculate the size of section need to be swizzled
+  //   numSwizzledThreadGroup = numTiles * tileDim
+  //
+  //   // Calculate the size of the side section
+  //   sideStart = numSwizzledThreadGroup.x * numSwizzledThreadGroup.y
+  //   sideWidth = numWorkgroups.x - numSwizzledThreadGroup.x
+  //   sideSize = sideWidth * numSwizzledThreadGroup.y
+  //
+  //   // Calculate the size of the bottom section
+  //   bottomStart = sideStart + sideSize
+  //   bottomHeight = numWorkgroups.y - numSwizzledThreadGroup.y
+  //
+  //   if (threadGroupFlatId >= bottomStart)
+  //     // Bottom tile
+  //     // Get new thread group ID for thread group in the bottom section
+  //     // Thread groups are reordered up->down then left->right
+  //     localThreadGroupFlatId = threadGroupFlatId - bottomStart
+  //     swizzledWorkgroupId.x = localThreadGroupFlatId / bottomHeight
+  //     swizzledWorkgroupId.y = (localThreadGroupFlatId % bottomHeight) + numSwizzledThreadGroup.y
+  //   else if (threadGroupFlatId >= sideStart)
+  //     // Side tile
+  //     // Get new thread group ID for thread group in the side section
+  //     // Thread groups are reordered left->right then up->down
+  //     localThreadGroupFlatId = threadGroupFlatId - sideStart
+  //     swizzledWorkgroupId.x = (localThreadGroupFlatId % sideWidth) + numSwizzledThreadGroup.x
+  //     swizzledWorkgroupId.y = localThreadGroupFlatId / sideWidth
+  //   else
+  //     // Morton tile
+  //     localThreadGroupFlatId = threadGroupFlatId % tileSize
+  //     // Extract to xy dimension based on Z-order curved
+  //     localThreadGroupId.x = Compact1By1Bits(tileBits, localThreadGroupFlatId)
+  //     localThreadGroupId.y = Compact1By1Bits(tileBits, localThreadGroupFlatId >> 1)
+  //     flatTileId = threadGroupFlatId / tileSize
+  //     swizzledWorkgroupId.x = (flatTileId % numTiles.x) * tileDim + localThreadGroupId.x
+  //     swizzledWorkgroupId.y = (flatTileId / numTiles.x) * tileDim + localThreadGroupId.y
+  //
+  //   // Finalize
+  //   swizzledWorkgroupId.z = nativeWorkgroupId.z
+  //
+  // Disable swizzle:
+  //   swizzledWorkgroupId = nativeWorkgroupId
+
+  BuilderBase builder(*m_context);
+
+  Type *ivec3Ty = FixedVectorType::get(Type::getInt32Ty(*m_context), 3);
+
+  auto func = m_module->getFunction(lgcName::SwizzleWorkgroupId);
+
+  func->setCallingConv(CallingConv::C);
+  func->addFnAttr(Attribute::NoUnwind);
+  func->addFnAttr(Attribute::AlwaysInline);
+  func->addFnAttr(Attribute::ReadNone);
+  func->setLinkage(GlobalValue::InternalLinkage);
+
+  auto argIt = func->arg_begin();
+
+  Value *numWorkgroups = argIt++;
+  numWorkgroups->setName("numWorkgroups");
+
+  Value *nativeWorkgroupId = argIt++;
+  nativeWorkgroupId->setName("nativeWorkgroupId");
+
+  static constexpr unsigned tileDims[] = {4, 8, 16};
+  static constexpr unsigned tileBits[] = {2, 3, 4};
+  static_assert((sizeof(tileDims) / sizeof(unsigned)) == (static_cast<unsigned>(ThreadGroupSwizzleMode::Count) - 1),
```

Comment:
nit: If you make `ThreadGroupSwizleMode` enumerable, you can use: `enumCount<ThreadGroupSwizzleMode`: https://github.com/GPUOpen-Drivers/llpc/blob/dev/lgc/interface/lgc/EnumIterator.h#L89